Applications/Spire/Source/KeyBindings/CancelKeyBindingsTableModel.cpp

```diff
@@ -0,0 +1,171 @@
+    static auto actions = std::vector<KeyBindings::CancelAction>({
+      KeyBindings::CancelAction::ALL,
+      KeyBindings::CancelAction::ALL_ASKS,
+      KeyBindings::CancelAction::ALL_BIDS,
+      KeyBindings::CancelAction::CLOSEST_ASK,
+      KeyBindings::CancelAction::CLOSEST_BID,
+      KeyBindings::CancelAction::FURTHEST_BID,
+      KeyBindings::CancelAction::FURTHEST_ASK,
+      KeyBindings::CancelAction::MOST_RECENT,
+      KeyBindings::CancelAction::MOST_RECENT_ASK,
+      KeyBindings::CancelAction::MOST_RECENT_BID,
+      KeyBindings::CancelAction::OLDEST,
+      KeyBindings::CancelAction::OLDEST_ASK,
+      KeyBindings::CancelAction::OLDEST_BID
+    });
+    return actions[index];
+  }
+
+  auto get_index(KeyBindings::CancelAction action) {
+    static auto indexes = std::unordered_map<KeyBindings::CancelAction, int>({
+      {KeyBindings::CancelAction::ALL, 0},
+      {KeyBindings::CancelAction::ALL_ASKS, 1},
+      {KeyBindings::CancelAction::ALL_BIDS, 2},
+      {KeyBindings::CancelAction::CLOSEST_ASK, 3},
+      {KeyBindings::CancelAction::CLOSEST_BID, 4},
+      {KeyBindings::CancelAction::FURTHEST_BID, 5},
+      {KeyBindings::CancelAction::FURTHEST_ASK, 6},
+      {KeyBindings::CancelAction::MOST_RECENT, 7},
+      {KeyBindings::CancelAction::MOST_RECENT_ASK, 8},
+      {KeyBindings::CancelAction::MOST_RECENT_BID, 9},
+      {KeyBindings::CancelAction::OLDEST, 10},
+      {KeyBindings::CancelAction::OLDEST_ASK, 11},
+      {KeyBindings::CancelAction::OLDEST_BID, 12}});
+    return indexes[action];
+  }
+}
+
+CancelKeyBindingsTableModel::CancelKeyBindingsTableModel(
+    const std::vector<KeyBindings::CancelActionBinding>& bindings,
+    QObject* parent)
+    : QAbstractTableModel(parent) {
+  for(auto i = 0; i < ROW_COUNT; ++i) {
+    m_key_bindings.push_back({{}, {}, get_action(i)});
+  }
+  set_key_bindings(bindings);
+}
+
+void CancelKeyBindingsTableModel::set_key_bindings(
+    const std::vector<KeyBindings::CancelActionBinding>& bindings) {
+  for(auto& new_binding : bindings) {
+    for(auto& existing_binding : m_key_bindings) {
+      if(existing_binding.m_sequence == new_binding.m_sequence &&
+          existing_binding.m_action != new_binding.m_action) {
+        existing_binding.m_sequence = QKeySequence();
+        break;
+      }
+    }
+    m_key_bindings[get_index(new_binding.m_action)].m_sequence =
+      new_binding.m_sequence;
+  }
+}
+
+KeyBindings::CancelAction CancelKeyBindingsTableModel::get_cancel_action(
+    int row) {
+  return get_action(row);
+}
+
+int CancelKeyBindingsTableModel::rowCount(const QModelIndex& parent) const {
+  return ROW_COUNT;
+}
+
+int CancelKeyBindingsTableModel::columnCount(const QModelIndex& parent) const {
+  return COLUMN_COUNT;
+}
+
+QVariant CancelKeyBindingsTableModel::data(const QModelIndex& index,
+    int role) const {
+  if(!index.isValid()) {
+    return QVariant();
+  }
+  if(role == Qt::DisplayRole) {
+    if(index.column() == 0) {
+      switch(index.row()) {
```

Comment:
Move this switch to a function (likely `constexpr const char*`) in the unnamed namespace, so the whole layout is in out place.